lib/commons/dom/create-grid.js

```diff
@@ -323,36 +323,66 @@ function findScrollRegionParent(vNode, parentVNode) {
  * @param {VirtualNode}
  */
 function addNodeToGrid(grid, vNode) {
-  const gridSize = constants.gridSize;
   vNode.clientRects.forEach(rect => {
-    if (rect.right <= 0 || rect.bottom <= 0) {
-      return;
-    }
     // save a reference to where this element is in the grid so we
     // can find it even if it's in a subgrid
     vNode._grid ??= grid;
-    const x = rect.left;
-    const y = rect.top;
+    const rectPosition = getGridPositionOfRect(rect);
+    grid.numCols = Math.max(grid.numCols ?? 0, rectPosition.endCol);
 
-    // "| 0" is a faster way to do Math.floor
-    // @see https://jsperf.com/math-floor-vs-math-round-vs-parseint/152
-    const startRow = (y / gridSize) | 0;
-    const startCol = (x / gridSize) | 0;
-    const endRow = ((y + rect.height) / gridSize) | 0;
-    const endCol = ((x + rect.width) / gridSize) | 0;
+    loopGridPosition(grid, rectPosition, gridCell => {
+      if (!gridCell.includes(vNode)) {
+        gridCell.push(vNode);
+      }
+    });
+  });
+}
 
-    grid.numCols = Math.max(grid.numCols ?? 0, endCol);
+export function loopGridPosition(grid, gridPosition, callback) {
+  const { startRow, endRow, startCol, endCol } = gridPosition;
+  loopNegativeIndexMatrix(grid.cells, startRow, endRow, row => {
+    loopNegativeIndexMatrix(row, startCol, endCol, callback);
+  });
+}
 
-    for (let row = startRow; row <= endRow; row++) {
-      grid.cells[row] = grid.cells[row] || [];
+// handle negative row/col values
+export function loopNegativeIndexMatrix(matrix, start, end, callback) {
+  matrix._negativeIndex ??= 0;
+  // Shift the array when start is negative
+  if (start < matrix._negativeIndex) {
+    for (let i = 0; i < matrix._negativeIndex - start; i++) {
+      matrix.splice(0, 0, []);
+    }
+    matrix._negativeIndex = start;
+  }
 
-      for (let col = startCol; col <= endCol; col++) {
-        grid.cells[row][col] = grid.cells[row][col] || [];
+  const startOffset = start - matrix._negativeIndex;
+  const endOffset = end - matrix._negativeIndex;
+  for (let index = startOffset; index <= endOffset; index++) {
+    matrix[index] ??= [];
+    callback(matrix[index], index + matrix._negativeIndex, index, matrix);
+  }
+}
 
-        if (!grid.cells[row][col].includes(vNode)) {
-          grid.cells[row][col].push(vNode);
-        }
-      }
-    }
-  });
+export function getGridPositionOfRect(
+  { top, left, bottom, right },
+  margin = 0
+) {
+  const { gridSize } = constants;
+  return {
+    startRow: Math.floor((top - margin) / gridSize),
+    startCol: Math.floor((left - margin) / gridSize),
+    endRow: Math.floor((bottom + margin - 1) / gridSize),
+    endCol: Math.floor((right + margin - 1) / gridSize)
+  };
+}
+
+export function getGridCellFromPoint({ cells, numCols }, { x, y }) {
+  const rowIndex = Math.floor(y / constants.gridSize);
+  const colIndex = Math.floor(x / constants.gridSize);
+  if (rowIndex >= cells.length || colIndex >= numCols) {
```

Comment:
This should take into account `_negativeIndex`. If the grid has 10 rows, but 9 of them are negative, a `rowIndex` of 2 should throw.

```suggestion
  if (rowIndex >= cells.length + cells._negativeIndex || colIndex >= numCols) {
```

Alternatively, we could set `numRows` like we do `numCols` as the largest `endRow` value so we don't have to use `length` of the  array.